src/loader/networks/sdex-chain.ts

```diff
@@ -28,7 +45,182 @@ export class SdexChain {
+      owner,
+      base,
+      quote,
+      poolIdx,
+    ]);
+    const returnData = await this.executeMulticall([{ target: this.config.query, callData }]);
+    const result = this.queryContract.interface.decodeFunctionResult('queryAmbientTokens', returnData);
+    return parseAmbientTokensResult(result);
+  }
+
+  public async queryRangeTokens(
+    owner: string,
+    base: string,
+    quote: string,
+    poolIdx: number,
+    lowTick: number,
+    upperTick: number,
+  ) {
+    this.validateInputs(owner, base, quote, poolIdx);
+
+    const callData = this.queryContract.interface.encodeFunctionData('queryRangeTokens', [
+      owner,
+      base,
+      quote,
+      poolIdx,
+      lowTick,
+      upperTick,
+    ]);
+
+    const returnData = await this.executeMulticall([{ target: this.config.query, callData }]);
+    const result = this.queryContract.interface.decodeFunctionResult('queryRangeTokens', returnData);
+    return parseRangeTokensResult(result);
+  }
+
+  private validateInputs(owner: string, base: string, quote: string, poolIdx: number) {
+    if (!owner || !base || !quote || poolIdx < 0) {
+      throw new Error('Invalid inputs');
+    }
+  }
+
+  private async getLPTokenBalance(user: string, baseToken: string): Promise<string> {
+    if (baseToken === ZeroAddress) {
+      const balance = await this.context.rpc.getBalance(user);
+      return balance.toString();
+    } else {
+      const LPContract = new ethers.Contract(
+        baseToken,
+        ['function balanceOf(address) view returns (uint256)'],
+        this.context.rpc,
+      );
+      const balance = await LPContract.balanceOf(user);
+      return balance.toString();
+    }
+  }
+
+  private async updateLPTokens(baseToken: string, quoteToken: string) {
+    if (!this.lpTokens.includes(baseToken)) {
+      this.lpTokens.push(baseToken);
+    }
+    if (!this.lpTokens.includes(quoteToken)) {
+      this.lpTokens.push(quoteToken);
+    }
+  }
+
+  public async getUpdatedLiquidity(user: string, base: string, quote: string, poolIdx: number) {
+    const { liquidityChanges } = await this.queryUserPositions(user);
+
+    const concentratedPositions = filterPositions(liquidityChanges, poolIdx, base, quote, PositionType.concentrated);
+    const ambientPositions = filterPositions(liquidityChanges, poolIdx, base, quote, PositionType.ambient);
+
+    if (concentratedPositions.length === 0 && ambientPositions.length === 0) {
+      return null;
+    }
+
+    await this.updateLPTokens(base, quote);
+
+    const ambientPositionResults = await Promise.all(
+      ambientPositions.map(async (userLiquidity) => {
+        const ambientTokens = await this.queryAmbientTokens(user, base, quote, poolIdx);
+        const [lpTokenBalanceBase, lpTokenBalanceQuote] = await Promise.all([
+          this.getLPTokenBalance(user, base),
+          this.getLPTokenBalance(user, quote),
+        ]);
```

Comment:
LP token is not `base` or `quote`, it's address retrievable from `this.query.queryPoolLpTokenAddress(base, quote, poolIdx)`

https://github.com/DistributedCollective/Sovryn-DEX/blob/7e10805c85f8f83db7a955a0a89dd91fedcea663/contracts/lens/SdexQuery.sol#L522